src/backend/gl/src/device.rs

```diff
@@ -468,16 +469,74 @@ pub(crate) unsafe fn set_sampler_info<SetParamFloat, SetParamFloatVec, SetParamI
 impl d::Device<B> for Device {
     unsafe fn allocate_memory(
         &self,
-        _mem_type: c::MemoryTypeId,
+        mem_type: c::MemoryTypeId,
         size: u64,
     ) -> Result<n::Memory, d::AllocationError> {
-        // TODO
-        Ok(n::Memory {
-            properties: memory::Properties::CPU_VISIBLE | memory::Properties::CPU_CACHED,
-            first_bound_buffer: Cell::new(None),
-            size,
-            emulate_map_allocation: RefCell::new(std::ptr::null_mut()),
-        })
+        if (1 << mem_type.0) & IMAGE_MEM_TYPE_MASK != 0 {
+            assert_ne!((1 << mem_type.0) & DEVICE_LOCAL_MASK, 0);
+            Ok(n::Memory {
+                properties: memory::Properties::DEVICE_LOCAL,
+                buffer: None,
+                target: 0,
+                size,
+                emulate_map_allocation: Cell::new(None),
+                map_flags: 0,
+            })
+        } else {
+            assert!(self.share.private_caps.buffer_role_change);
+
+            let target = if (1 << mem_type.0) & INDEX_MEM_TYPE_MASK != 0 {
```

Comment:
similarly, could we compare directly instead?

Reply:
There's more than one type of index memory, so the intention here is to check whether the given memory type is any of the index memory types.  Is there a clearer way to write this?